doc/source/indexing.rst

```diff
@@ -667,6 +667,94 @@ For getting *multiple* indexers, using ``.get_indexer``
   dfd.iloc[[0, 2], dfd.columns.get_indexer(['A', 'B'])]
 
 
+.. _indexing.deprecate_loc_reindex_listlike:
+
+Using loc with missing keys in a list is Deprecated
+---------------------------------------------------
+
+.. warning::
+
+   Starting in 0.21.0, using ``.loc`` with a list-like containing missing key, is deprecated, in favor of ``.reindex``.
+
+In prior versions, using ``.loc[list-of-keys]`` would work as long as *at least 1* of the keys was found (otherwise it
+would raise a ``KeyError``). This behavior is deprecated and will show a warning message pointing to this section. The
+recommeded alternative is to use ``.reindex()``.
+
+For example.
+
+.. ipython:: python
+
+   s = Series([1, 2, 3])
+   s
+
+Selection with all keys found is unchanged.
+
+.. ipython:: python
+
+   s.loc[[1, 2]]
+
+Previous Behavior
+
+.. code-block:: ipython
+
+
+   In [4]: s.loc[[1, 2, 3]]
+   Out[4]:
+   1    2.0
+   2    3.0
+   3    NaN
+   dtype: float64
+
+
+Current Behavior
+
+   In [4]: s.loc[[1, 2, 3]]
+   Passing list-likes to .loc with any non-matching elements will raise
+   KeyError in the future, you can use .reindex() as an alternative.
+
+   See the documentation here:
+   http://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
+
+   Out[4]:
+   1    2.0
+   2    3.0
+   3    NaN
+   dtype: float64
+
+
+Reindexing
+~~~~~~~~~~
+
+The idiomatic way to achieve selecting potentially not-found elmenents is via ``.reindex()``. See also the section on :ref:`reindexing <basics.reindexing>`.
+
+.. ipython:: python
+
+  s.reindex([1, 2, 3])
+
+Alternatively, if you want to select only *valid* keys, the following is idiomatic; furthermore this is more efficient, and is guaranteed to preserve the dtype of the selection.
+
+.. ipython:: python
+
+   keys = [1, 2, 3]
+   s.loc[s.index & keys]
+
+Having a duplicated index will raise for a ``.reindex()``:
+
+.. ipython:: python
+
+   s = pd.Series(np.arange(4), index=['a', 'a', 'b', 'c'])
+
+.. code-block:: python
+
+   In [17]: s.reindex(['c', 'd'])
+   ValueError: cannot reindex from a duplicate axis
+
+The idiomatic expression again allows this operation to proceed
+
+.. ipython:: python
+
+   s.loc[s.index & ['c', 'd']]
```

Comment:
I would personally go for the full `intersection` notation, that is IMO more readable / less confusing (since the typical use of `&` is for boolean comparisons, that's what most users will know)

Reply:
making ``.loc`` handle the case of duplicate indexers is non-trivial and certainly complicated things (the code path is still there, may have to remove in the future).

Reply:
Repeating my comment:

- I would use `intersection` instead of `&`
- this is not a correct example (the alternative you provide does not give the same result as the deprecated one. As I say above I am not sure *if* there is actually an alternative, but if there is not, we should just remove this example)

Reply:
Can you answer to the second point? (as according to the diff, you only addressed my first point) BTW, you don't necessarily need to agree with me / update the code according to my comment, but in that case would be nice to get your view.

So my point is: it is not replicating the previous behaviour of `.loc` (the result is different), and therefore I would either just remove the example or explain that with duplicates there is no alternative, but you can get the the existing labels with this method

Reply:
Can you answer this comment?

Reply:
this is already updated, I think you are looking at an old diff

Reply:
OK, I suppose I missed the ´.reindex(labels)´ you added in `s.loc[s.index.intersection(labels)].reindex(labels)` (but a separate commit afer the comment above / an answer would have been helpful to notice it).

Anyhow, on the actual topic: I am still not fully convinced this is a good example. By adding the `reindex` it indeed now gives the same output as `loc` before for the example, but that is only because you didn't select one of the duplicate keys. If you do that, the proposed solution fails (both the `intersection` as the `reindex` fail in that case):

```
In [36]: s = pd.Series(np.arange(4), index=['a', 'a', 'b', 'c'])

In [39]: labels = ['c', 'd']

In [40]: s.loc[s.index.intersection(labels)].reindex(labels)
Out[40]: 
c    3.0
d    NaN
dtype: float64

In [41]: labels = ['a', 'd']

In [42]: s.loc[s.index.intersection(labels)].reindex(labels)
...
ValueError: cannot reindex from a duplicate axis

```

Reply:
yes and it should fail, that is the point. reindexing across duplicate keys is way too magical. yes I built that into ``.loc`` a while back, but it is being removed by this change. This is a good thing.

Reply:
it still is technically in ``.loc`` (you will get a warning though)

Reply:
I am not saying it should not fail. I totally agree that this fails. 
What I am trying to say is that it is therefore not a very robust example, and if we do not have a good alternative for this, we should just leave it out IMO.

Reply:
ok, tell you what I will add on [41] & [42] as well to show this too.

Reply:
pushed

Reply:
OK!